### ISSUE TITLE
High CPU usage on UXG Lite with Intrusion Prevention enabled

### ISSUE THREAD

**Issue by dmutsaers** [fabianishere/udm-iptv#361]
### Diagnostic Information

=== Configuration ===
WAN Interface: eth1
WAN VLAN: 4 (dev iptv)
WAN DHCP: true (options "-O staticroutes -V IPTV_RG")
WAN Ranges: 10.0.0.0/8  10.252.0.0/16  10.253.0.0/16  217.166.0.0/16
LAN Interfaces: eth0.8
IGMP Proxy quickleave disabled: false
IGMP Proxy debug: false
=== IP Link and Route ===
17: iptv@eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet 10.252.214.207/16 scope global iptv
       valid_lft forever preferred_lft forever
10.0.0.0/8 via 10.252.0.1 metric 217
10.252.0.0/16 proto kernel scope link src 10.252.214.207
=== Service Logs ===
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]:         group time 208
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]:         version: 2
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]: group 224.2.2.2 exclude mode
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]:         group time 204
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]:         version: 2
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]: interface eth0.8:192.168.8.1
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]: group 224.0.251.124 exclude mode
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]:         group time 260
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]:         version: 2
Dec 10 10:01:49 Next-GenGatewayLite udm-iptvd[10937]: group 224.0.251.109 exclude mode

### Describe the Bug

If the latest Unifi firmware (4.1.2/4.1.3) has been installed on the UXG Lite/Max/Pro/Enterprise the CPU usage maxes out to 100% when an IPTV stream is active using udm-iptv. IPTV becomes unusable. It seems the ulogd process uses all CPU capacity.
Older firmware doesn't seem to suffer from this issue.

[High CPU usage](https://community.ui.com/questions/High-CPU-usage/7b178f8c-6b2b-4fc4-835b-7b129ec52958) (Unifi forum)

[Unifi Gateways firmware 4.1.2 + 4.1.3 incompatible with udm-iptv (UXG Lite/Max/Pro/Enterprise) #358](https://github.com/fabianishere/udm-iptv/discussions/358)

### Expected Behavior

Installing udm-iptv shouldn't cause 100% CPU usage on UXG Lite/Max/Pro/Enterprise.

**Comment by dmutsaers** [comment]
Unifi's response: "We cannot assure stability on modified OS's, but I understand why you're adjusting the OS."

**Comment by dmutsaers** [comment]
At least one user seems to have no problems with the latest firmware on a Cloud Gateway.

"Yes, I am running udm-iptv on my Cloud Gateway Ultra with v4.1.9
I see no issues personally (KPN)"

https://community.ui.com/releases/UniFi-OS-Cloud-Gateways-4-1-9/10ac13d7-188a-4bc9-b83e-53239a9ca0e9#comment/1f8dd580-6996-4837-a195-5900e323eea7

**Comment by dmutsaers** [comment]
[UniFi Gateways 4.1.8](https://community.ui.com/releases/UniFi-Gateways-4-1-8/5f9fefe3-3557-4f62-b754-fb4ff7980d5c) has been released, IPTV still not working.

**Comment by spiralshapeturtle** [comment]
@dmutsaers which process at the UI device is consuming the CPU? Could you post an "top" screenshot when the issues is present?

**Comment by dmutsaers** [comment]
I think disabling Intrusion Prevention "fixes" the issue:
![image](https://github.com/user-attachments/assets/5e7b225e-fc38-4af2-a0f4-0acae7f5f57f)

**Comment by latereaction** [comment]
> I think disabling Intrusion Prevention "fixes" the issue:
> 
> ![image](https://private-user-images.githubusercontent.com/45763846/401170107-5e7b225e-fc38-4af2-a0f4-0acae7f5f57f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNDQ4OTAsIm5iZiI6MTczNjM0NDU5MCwicGF0aCI6Ii80NTc2Mzg0Ni80MDExNzAxMDctNWU3YjIyNWUtZmMzOC00YWYyLWEwZjQtMGFjYWU3ZjVmNTdmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDEzNTYzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjZTcyOWE1ZjVjY2Y3Y2Q2ZmEyOTI1MTRiYzBiMWRhMjc3OGEzMTM2MDUzNzMyMzUwZjRiMGUyYTI4NzMzYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yL2ziivdr20wk0fYIRGWXVnyxAgcJE57aX5alMI3dOc)

Not for me - I don't have IPS enabled.

**Comment by spiralshapeturtle** [comment]
@dmutsaers ksoftirqd (packetforwarder if I recall well) is consuming the CPU you mean in the second screenshot? In the UI topic thy talk about ulog as consumer, could you elaborate?

**Comment by spiralshapeturtle** [comment]
> > I think disabling Intrusion Prevention "fixes" the issue:
> > ![image](https://private-user-images.githubusercontent.com/45763846/401170107-5e7b225e-fc38-4af2-a0f4-0acae7f5f57f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzNDQ4OTAsIm5iZiI6MTczNjM0NDU5MCwicGF0aCI6Ii80NTc2Mzg0Ni80MDExNzAxMDctNWU3YjIyNWUtZmMzOC00YWYyLWEwZjQtMGFjYWU3ZjVmNTdmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMTA4VDEzNTYzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjZTcyOWE1ZjVjY2Y3Y2Q2ZmEyOTI1MTRiYzBiMWRhMjc3OGEzMTM2MDUzNzMyMzUwZjRiMGUyYTI4NzMzYTQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yL2ziivdr20wk0fYIRGWXVnyxAgcJE57aX5alMI3dOc)
> 
> Not for me - I don't have IPS enabled.

Which proces is eating your CPU?

**Comment by dmutsaers** [comment]
Strange, just by enabling or disabling "Intrusion Prevention" I can create or solve the issue.

**Comment by spiralshapeturtle** [comment]
> Strange, just by enabling or disabling "Intrusion Prevention" I can create or solve the issue.

I dont think that is strange, what I think is the multicast traffic or the inspection of this content is pushed over the CPU instead of hardware offloading. So you just "use" the CPU when you enable the inspection it? I assume in every other speedtest.net test the CPU remains very low due to the hardware offloading?

**Comment by dmutsaers** [comment]
> I dont think that is strange, what I think is the multicast traffic or the inspection of this content is pushed over the CPU instead of hardware offloading. So you just "use" the CPU when you enable the inspection it? I assume in every other speedtest.net test the CPU remains very low due to the hardware offloading?

That may be true, it never happened before in the 4.0.x branch. A speedtest does not stress the CPU.

**Comment by latereaction** [comment]
> > @dmutsaers ksoftirqd (packetforwarder if I recall well) is consuming the CPU you mean in the second screenshot? In the UI topic thy talk about ulog as consumer, could you elaborate?
> 
> I've never checked which process was consuming all CPU. The first time I looked at the processes with TOP was after upgrading to 4.1.8. today (at your request). I noticed the ksoftirq and ubnt-idsips-dae processes we're consuming all CPU on my UXG-Lite.

I monitored it for quite some time on 4.1.2 and for me it was the ULOGD process that was consuming all CPU.

**Comment by spiralshapeturtle** [comment]
> > I dont think that is strange, what I think is the multicast traffic or the inspection of this content is pushed over the CPU instead of hardware offloading. So you just "use" the CPU when you enable the inspection it? I assume in every other speedtest.net test the CPU remains very low due to the hardware offloading?
> 
> That may be true, it never happened before in the 4.0.x branch. A speedtest does not stress the CPU.

Hardware offloading is introduced in 4.1, so yes the packetforwarding did change from 4.0 to 4.1

**Comment by spiralshapeturtle** [comment]
```
# Disable quickleave for igmpproxy
IPTV_IGMPPROXY_DISABLE_QUICKLEAVE="false"
# Enable debugging for igmpproxy
IPTV_IGMPPROXY_DEBUG="false"
IPTV_IGMPPROXY_PROGRAM="igmpproxy"
```

Does igmpproxy make a difference, instead of the improxy?

**Comment by dmutsaers** [comment]
That makes a lot of difference!
(IGMPPROXY isn't downloaded/installed by default, I had to "apt install")

TOP (IGMPPROXY, Intrusion Prevention disabled)
![Snipaste_2025-01-08_15-57-43](https://github.com/user-attachments/assets/67a9f939-b1f3-4ebb-a033-09ea9020e5c0)
TOP (IGMPPROXY, Intrusion Prevention enabled)
![Snipaste_2025-01-08_15-51-01](https://github.com/user-attachments/assets/3ca233ca-2b2f-4d51-a5cb-5482f7ddca2b)

**Comment by dmutsaers** [comment]
I've been experimenting with IGMPPROXY, IMPROXY and Intrusion Prevention. I have best results with Intrusion Prevention disabled. I could enable Intrusion Prevention with IGMPPROX, the IPTV stream is stable, but it still stresses the CPU a lot. Enabling Intrusion Prevention with IMPROXY is not possible.

**Comment by rogierlommers** [comment]
> I've been experimenting with IGMPPROXY, IMPROXY and Intrusion Prevention. I have best results with Intrusion Prevention disabled. I could enable Intrusion Prevention with IGMPPROX, the IPTV stream is stable, but it still stresses the CPU a lot. Enabling Intrusion Prevention with IMPROXY is not possible.

Nice learnings! Thanks for posting!

**Comment by spiralshapeturtle** [comment]
> I've been experimenting with IGMPPROXY, IMPROXY and Intrusion Prevention. I have best results with Intrusion Prevention disabled. I could enable Intrusion Prevention with IGMPPROXY, the IPTV stream is stable, but it still stresses the CPU a lot. Enabling Intrusion Prevention with IMPROXY is not possible.

Which unit do you have, this is a UXG-Max 4.1.8 with the fabian script, inspection on, IGMPPROXY (IMPROXY stops after a few seconds), debug logs off, STB vlan IGMP snooping enabled. 

I'm not seeing high CPU?

<img width="910" alt="image" src="https://github.com/user-attachments/assets/f7cbe67a-a3fb-4993-9f88-9117322868bd" />

**Comment by dmutsaers** [comment]
> Which unit do you have, this is a UXG-Max 4.1.8 with the fabian script, inspection on, IGMPPROXY (IMPROXY stops after a few seconds), debug logs off, STB vlan IGMP snooping enabled.
> 
> I'm not seeing high CPU?

I'm using an UXG-Lite. High CPU is only observed with an active multicast IGMP IPTV stream with Intrusion Prevention enabled.

**Comment by dmutsaers** [comment]
> Not for me - I don't have IPS enabled.

Just for fun, could you enable Intrusion Prevention, and after enabling Intrusion Prevention disable Intrusion Prevention again?

**Comment by spiralshapeturtle** [comment]
> > Which unit do you have, this is a UXG-Max 4.1.8 with the fabian script, inspection on, IGMPPROXY (IMPROXY stops after a few seconds), debug logs off, STB vlan IGMP snooping enabled.
> > I'm not seeing high CPU?
> 
> I'm using an UXG-Lite. High CPU usage is only observed with an active multicast IGMP IPTV stream and Intrusion Prevention enabled.

Of course my test was with IPTV (KPN) enabled/streaming ;-) Could you change the first post/title that the issue might only occurs to the UXG-LITE hardware? It's a bit misleading now for others who search via google.

**Comment by latereaction** [comment]
> > Not for me - I don't have IPS enabled.
> 
> Just for fun, could you enable Intrusion Prevention, and after enabling Intrusion Prevention disable Intrusion Prevention again?

Ill include that test when i upgrade tonight.

**Comment by spiralshapeturtle** [comment]
@dmutsaers you might want to upgrade your hardware, the UXG-LITE got a different chipset then the UXG-Max.

IPS off

<img width="871" alt="image" src="https://github.com/user-attachments/assets/1ef8f498-dc8c-40e6-a044-bdf1e1853eb9" />

IPS enabled

<img width="768" alt="image" src="https://github.com/user-attachments/assets/859c8988-8c18-4556-84cf-556208bfba26" />

**Comment by spiralshapeturtle** [comment]
> > > @dmutsaers ksoftirqd (packetforwarder if I recall well) is consuming the CPU you mean in the second screenshot? In the UI topic thy talk about ulog as consumer, could you elaborate?
> > 
> > 
> > I've never checked which process was consuming all CPU. The first time I looked at the processes with TOP was after upgrading to 4.1.8. today (at your request). I noticed the ksoftirq and ubnt-idsips-dae processes we're consuming all CPU on my UXG-Lite.
> 
> I monitored it for quite some time on 4.1.2 and for me it was the ULOGD process that was consuming all CPU.

Just checking, UXG-LITE hardware?

**Comment by dmutsaers** [comment]
> @dmutsaers you might want to upgrade your hardware, the UXG-LITE got a different chipset then the UXG-Max.
> 
> Below UXG-Max
> 
> IPS off
> IPS enabled

Your results make me wonder there is something really wrong with the UXG-Lite.
A dual-core ARM Cortex-A53 CPU running at 1 GHz and a quad-core ARM Cortex-A53 CPU running at 1.5 GHz shouldn't make such a big difference, should it?

**Comment by dmutsaers** [comment]
UXG-Lite, IGMPPROXY, IPS off
![image](https://github.com/user-attachments/assets/701fb681-9132-4ec3-8a89-405f9c7f6af4)

**Comment by spiralshapeturtle** [comment]
> > @dmutsaers you might want to upgrade your hardware, the UXG-LITE got a different chipset then the UXG-Max.
> > Below UXG-Max
> > IPS off
> > IPS enabled
> 
> Your results make me wonder there is something really wrong with the UXG-Lite. A dual-core ARM Cortex-A53 CPU running at 1 GHz and a quad-core ARM Cortex-A53 CPU running at 1.5 GHz shouldn't make such a big difference, should it?

The forwarding chip is another one which makes it a total different device. This is not related to the CPU capacity. In my screenshots there is no CPU load for forwarding. Somehow the UXG-Max does handle this traffic differently than the UXG-Lite.

More about chipsets overhere.
https://community.ui.com/questions/UniFi-Network-Product-Chipset-Matrix-and-Birth/feb69f12-321d-4102-a5dd-bbab81d08483#answer/6c22a9f9-dbdb-4523-9c5e-564f362af9f5

**Comment by latereaction** [comment]
OK so I have some interesting/confusing news - I re-upgraded to 4.1.8 last night, reinstalled udm-iptv and switched it in the config to igmpproxy(this is all currently with IPS off) and I am seeing completely normal CPU usage with active IPTV streams. Curiosity then got the better of me and i switched back to improxy, as igmpproxy seems to use a little more RAM. Still seeing completely normal CPU usage. here is my usage graph from UNA for last night when the IPTV would have been on, and overnight to this morning:
![image](https://github.com/user-attachments/assets/ac1dbd40-0f56-403f-b248-d7645f342771)

I haven't changed anything in my config other than keeping UNA early access up to date, but I don't see how that could have an effect (happy to be corrected).

I ran out of time to try it with IPS enabled, so that will be todays task, however I'm not expecting great things as the UXG Lite struggles really with it anyway.

**Comment by dmutsaers** [comment]
@latereaction A bit similar as my experiences, it all crumbles down if you enable IPS.

**Comment by latereaction** [comment]
> @latereaction A bit similar as my experiences, it all crumbles down if you enable IPS.

Yeah previously it would have been obvious as the CPU would be pinned at 100% for long periods during IPTV streams.

I think even though Ubiquiti advertise this device as IPS capable - i really think its at its limit. I ended up disabling it pretty sharp after getting the device and seeing the poor performance. I do run Ad blocker on some of my VLANS and also use country blocking, but that's it.

Its a little frustrating that we don't know the root cause of this though, as until Ubiquiti fix their broken (for us) IPTV implementation we are at the behest of the performance of this add-on on all future firmwares.
